manager/container_test.go

```diff
@@ -226,3 +231,66 @@ func TestUpdateNvidiaStats(t *testing.T) {
 	assert.NotNil(t, err)
 	assert.Equal(t, info.ContainerStats{}, stats)
 }
+
+func TestOnDemandHousekeeping(t *testing.T) {
+	statsList := itest.GenerateRandomStats(1, 4, 1*time.Second)
+	stats := statsList[0]
+
+	cd, mockHandler, memoryCache, fakeClock := newTestContainerData(t)
+	mockHandler.On("GetStats").Return(stats, nil)
+	defer cd.Stop()
+
+	// 0 seconds should always trigger an update
+	go cd.OnDemandHousekeeping(0 * time.Second)
+	cd.housekeepingTick(fakeClock.NewTimer(time.Minute).C(), testLongHousekeeping)
+
+	fakeClock.Step(2 * time.Second)
+
+	// This should return without requiring a housekeepingTick because stats have been updated recently enough
+	cd.OnDemandHousekeeping(3 * time.Second)
+
+	go cd.OnDemandHousekeeping(1 * time.Second)
+	cd.housekeepingTick(fakeClock.NewTimer(time.Minute).C(), testLongHousekeeping)
+
+	checkNumStats(t, memoryCache, 2)
+	mockHandler.AssertExpectations(t)
+}
+
+func TestConcurrentOnDemandHousekeeping(t *testing.T) {
+	statsList := itest.GenerateRandomStats(1, 4, 1*time.Second)
+	stats := statsList[0]
+
+	cd, mockHandler, memoryCache, fakeClock := newTestContainerData(t)
+	mockHandler.On("GetStats").Return(stats, nil)
+	defer cd.Stop()
+
+	numConcurrentCalls := 5
+	var waitForHousekeeping sync.WaitGroup
+	waitForHousekeeping.Add(numConcurrentCalls)
+	onDemandCache := make(chan chan struct{}, numConcurrentCalls)
+	for i := 0; i < numConcurrentCalls; i++ {
+		go func() {
+			cd.OnDemandHousekeeping(0 * time.Second)
+			waitForHousekeeping.Done()
+		}()
+		// Wait for work to be queued
+		onDemandCache <- <-cd.onDemandChan
```

Comment:
nit: I'd prefer a slice for the cache. It makes the requeue more straightforward.

Reply:
yeah, but the double arrow looks really cool.  Done